### PR TITLE
Use interpolated string handler for Contract.ThrowIf methods

### PR DESCRIPTION
**Pull request by Youssef1313**
Closes https://github.com/dotnet/roslyn/issues/58165

cc @333fred @sharwell @CyrusNajmabadi 

I'm not sure if I'm doing it right

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.cs [hunk 1]
```diff
@@ -6,11 +6,50 @@
...
+        private struct ThrowIfTrueInterpolatedStringHandler
+        {
+            private readonly StringBuilder? _stringBuilder;
+
+#pragma warning disable IDE0060 // Remove unused parameter - https://github.com/dotnet/roslyn/issues/58168
+            public ThrowIfTrueInterpolatedStringHandler(int literalLength, int formattedCount, bool condition)
+#pragma warning restore IDE0060 // Remove unused parameter
+            {
+                _stringBuilder = condition ? new StringBuilder(capacity: literalLength) : null;
+            }
```
Comment (333fred) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.cs:25]:
```suggestion
            public ThrowIfTrueInterpolatedStringHandler(int literalLength, int formattedCount, bool condition, out bool success)
#pragma warning restore IDE0060 // Remove unused parameter
            {
                _stringBuilder = condition ? new StringBuilder(capacity: literalLength) : null;
                success = condition;
            }
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.cs [hunk 2]
```diff
@@ -6,11 +6,50 @@
...
+        private struct ThrowIfFalseInterpolatedStringHandler
+        {
+            private readonly StringBuilder? _stringBuilder;
+
+#pragma warning disable IDE0060 // Remove unused parameter - https://github.com/dotnet/roslyn/issues/58168
+            public ThrowIfFalseInterpolatedStringHandler(int literalLength, int formattedCount, bool condition)
+#pragma warning restore IDE0060 // Remove unused parameter
+            {
+                _stringBuilder = condition ? null : new StringBuilder(capacity: literalLength);
+            }
```
Comment (333fred) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.cs:44]:
```suggestion
            public ThrowIfFalseInterpolatedStringHandler(int literalLength, int formattedCount, bool condition, out bool success)
#pragma warning restore IDE0060 // Remove unused parameter
            {
                _stringBuilder = condition ? null : new StringBuilder(capacity: literalLength);
                success = !condition;
            }
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.cs [hunk 3]
```diff
@@ -48,7 +87,7 @@ internal static class Contract
         /// all builds
         /// </summary>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static void ThrowIfNull<T>([NotNull] T value, string message, [CallerLineNumber] int lineNumber = 0)
+        public static void ThrowIfNull<T>([NotNull] T value, string message, [CallerLineNumber] int lineNumber = 0) // TODO: Interpolated string handler.
```
Comment (333fred) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.cs:90]: You'll want a `ThrowIfNullHandler<T>` that takes a `T` and `out`'s a `bool` for this one.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.cs [hunk 4]
```diff
@@ -74,11 +113,11 @@ public static void ThrowIfFalse([DoesNotReturnIf(parameterValue: false)] bool co
         /// in all builds
         /// </summary>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static void ThrowIfFalse([DoesNotReturnIf(parameterValue: false)] bool condition, string message, [CallerLineNumber] int lineNumber = 0)
+        public static void ThrowIfFalse([DoesNotReturnIf(parameterValue: false)] bool condition, [InterpolatedStringHandlerArgument("condition")] ThrowIfFalseInterpolatedStringHandler message, [CallerLineNumber] int lineNumber = 0)
```
Comment (333fred) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.cs:116]: We should add these as overloads, not replacements. Overload resolution will pick the handler for interpolated constants.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.cs [hunk 5]
```diff
@@ -6,11 +6,50 @@
...
+#pragma warning restore IDE0060 // Remove unused parameter
+            {
+                _stringBuilder = condition ? new StringBuilder(capacity: literalLength) : null;
+            }
+
+            public void AppendLiteral(string value) => _stringBuilder?.Append(value);
+
+            public void AppendFormatted<T>(T value) => _stringBuilder?.Append(value?.ToString());
+
+            public string GetFormattedText() => _stringBuilder?.ToString() ?? string.Empty;
```
Comment (333fred) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.cs:31]: For these, I would actually assert that `_stringBuilder` isn't null. The API is being used wrong if it is.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.InterpolatedStringHandlers.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+            public ThrowIfNullInterpolatedStringHandler(int literalLength, int formattedCount, T? value, out bool success)
+#pragma warning restore IDE0060 // Remove unused parameter
+            {
+                _stringBuilder = value is null ? new StringBuilder(capacity: literalLength) : null;
+                success = value is null;
+            }
+
+            public void AppendLiteral(string value) => _stringBuilder?.Append(value);
+
+            public void AppendFormatted<T2>(T2 value) => _stringBuilder?.Append(value?.ToString());
```
Comment (Youssef1313) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.InterpolatedStringHandlers.cs:67]: Couldn't find good names for the type parameters.

File: src/Compilers/Core/Portable/InternalUtilities/InterpolatedStringHandlerArgumentAttribute.cs [hunk 7]
```diff
@@ -0,0 +1,23 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Runtime.CompilerServices
```
Comment (333fred) [inline src/Compilers/Core/Portable/InternalUtilities/InterpolatedStringHandlerArgumentAttribute.cs:5]: These files should probably only be included if we're not on .NET 6 or greater, using the `#if NET6_0_OR_GREATER` defines.

**Review by 333fred — APPROVED** [review]

**Comment by Youssef1313** [comment]
@333fred Anything left here?

**Comment by 333fred** [comment]
@sharwell @dotnet/roslyn-ide for review.

**Comment by Youssef1313** [comment]
@sharwell for review.

**Comment by Youssef1313** [comment]
@sharwell Can you review please? Thanks!

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.InterpolatedStringHandlers.cs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            {
+                _stringBuilder = condition ? new StringBuilder(capacity: literalLength) : null;
+                success = condition;
+            }
+
+            public void AppendLiteral(string value) => _stringBuilder?.Append(value);
+
+            public void AppendFormatted<T>(T value) => _stringBuilder?.Append(value?.ToString());
+
+            public string GetFormattedText() => _stringBuilder!.ToString();
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.InterpolatedStringHandlers.cs:29]: I lean towards treating this like we treat nullability for implementations of `IEnumerator<T>`. The nullability annotation represents the _valid_ object state in order to minimize the number of suppressions.

```suggestion
            private readonly StringBuilder _stringBuilder;

#pragma warning disable IDE0060 // Remove unused parameter - https://github.com/dotnet/roslyn/issues/58168
            public ThrowIfTrueInterpolatedStringHandler(int literalLength, int formattedCount, bool condition, out bool success)
#pragma warning restore IDE0060 // Remove unused parameter
            {
                _stringBuilder = condition ? new StringBuilder(capacity: literalLength) : null!;
                success = condition;
            }

            public void AppendLiteral(string value) => _stringBuilder.Append(value);

            public void AppendFormatted<T>(T value) => _stringBuilder.Append(value?.ToString());

            public string GetFormattedText() => _stringBuilder.ToString();
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.InterpolatedStringHandlers.cs [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+            {
+                _stringBuilder = condition ? null : new StringBuilder(capacity: literalLength);
+                success = !condition;
+            }
+
+            public void AppendLiteral(string value) => _stringBuilder?.Append(value);
+
+            public void AppendFormatted<T>(T value) => _stringBuilder?.Append(value?.ToString());
+
+            public string GetFormattedText() => _stringBuilder!.ToString();
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.InterpolatedStringHandlers.cs:49]:
```suggestion
            private readonly StringBuilder _stringBuilder;

#pragma warning disable IDE0060 // Remove unused parameter - https://github.com/dotnet/roslyn/issues/58168
            public ThrowIfFalseInterpolatedStringHandler(int literalLength, int formattedCount, bool condition, out bool success)
#pragma warning restore IDE0060 // Remove unused parameter
            {
                _stringBuilder = condition ? null! : new StringBuilder(capacity: literalLength);
                success = !condition;
            }

            public void AppendLiteral(string value) => _stringBuilder.Append(value);

            public void AppendFormatted<T>(T value) => _stringBuilder.Append(value?.ToString());

            public string GetFormattedText() => _stringBuilder.ToString();
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.InterpolatedStringHandlers.cs [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+            {
+                _stringBuilder = value is null ? new StringBuilder(capacity: literalLength) : null;
+                success = value is null;
+            }
+
+            public void AppendLiteral(string value) => _stringBuilder?.Append(value);
+
+            public void AppendFormatted<T2>(T2 value) => _stringBuilder?.Append(value?.ToString());
+
+            public string GetFormattedText() => _stringBuilder!.ToString();
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.InterpolatedStringHandlers.cs:69]:
```suggestion
            private readonly StringBuilder _stringBuilder;

#pragma warning disable IDE0060 // Remove unused parameter - https://github.com/dotnet/roslyn/issues/58168
            public ThrowIfNullInterpolatedStringHandler(int literalLength, int formattedCount, T? value, out bool success)
#pragma warning restore IDE0060 // Remove unused parameter
            {
                _stringBuilder = value is null ? new StringBuilder(capacity: literalLength) : null!;
                success = value is null;
            }

            public void AppendLiteral(string value) => _stringBuilder.Append(value);

            public void AppendFormatted<T2>(T2 value) => _stringBuilder.Append(value?.ToString());

            public string GetFormattedText() => _stringBuilder.ToString();
```

**Review by sharwell — APPROVED** [review]

**Review by sharwell — APPROVED** [review]

**Review by 333fred — COMMENTED** [review]
@dotnet/roslyn-compiler for a second review of the compiler files.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.InterpolatedStringHandlers.cs [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace Roslyn.Utilities
+{
+    internal static partial class Contract
+    {
+        [InterpolatedStringHandler]
+        public struct ThrowIfTrueInterpolatedStringHandler
```
Comment (jaredpar) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.InterpolatedStringHandlers.cs:13]:
```suggestion
        public readonly struct ThrowIfTrueInterpolatedStringHandler
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.InterpolatedStringHandlers.cs [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+            public string GetFormattedText() => _stringBuilder.ToString();
+        }
+
+        [InterpolatedStringHandler]
+        public struct ThrowIfFalseInterpolatedStringHandler
+        {
+            private readonly StringBuilder _stringBuilder;
+
+#pragma warning disable IDE0060 // Remove unused parameter - https://github.com/dotnet/roslyn/issues/58168
```
Comment (jaredpar) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.InterpolatedStringHandlers.cs:37]: Consider disabling at the file level. Feel like the constant disable / enable in this file is just making code harder to read.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.InterpolatedStringHandlers.cs [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+
+            public void AppendLiteral(string value) => _stringBuilder.Append(value);
+
+            public void AppendFormatted<T>(T value) => _stringBuilder.Append(value?.ToString());
+
+            public string GetFormattedText() => _stringBuilder.ToString();
+        }
+
+        [InterpolatedStringHandler]
+        public struct ThrowIfFalseInterpolatedStringHandler
```
Comment (jaredpar) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.InterpolatedStringHandlers.cs:33]:
```suggestion
        public readonly struct ThrowIfFalseInterpolatedStringHandler
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.InterpolatedStringHandlers.cs [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+            public void AppendLiteral(string value) => _stringBuilder.Append(value);
+
+            public void AppendFormatted<T>(T value) => _stringBuilder.Append(value?.ToString());
+
+            public string GetFormattedText() => _stringBuilder.ToString();
+        }
+
+        [InterpolatedStringHandler]
+        public struct ThrowIfNullInterpolatedStringHandler<T>
```
Comment (jaredpar) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.InterpolatedStringHandlers.cs:53]:
```suggestion
        public readonly struct ThrowIfNullInterpolatedStringHandler<T>
```

**Comment by 333fred** [comment]
@jaredpar any more feedback?

**Comment by Youssef1313** [comment]
@jaredpar Can you take a look please? Thanks!

**Review by jaredpar — APPROVED** [review]

**Pull request merged**